src/azul/plugins/repository/canned/__init__.py

```diff
@@ -69,7 +72,104 @@
 log = logging.getLogger(__name__)
 
 
-class CannedSourceRef(SourceRef[SimpleSourceSpec, 'CannedSourceRef']):
+@attrs.frozen(kw_only=True)
+class CannedSourceSpec(SourceSpec):
+    url: str
+    ref: str
+    path: tuple[str]
+
+    @classmethod
+    def parse(cls, spec: str) -> Self:
+        """
+        Construct an instance from its string representation, using the syntax
+        '{repo-url}::{ref}::{path}::{prefix}' where:
+        {repo-url} is the HTTPS URL of a GitHub repository's `.git` file. It
+                   must match the format https://github.com/USER/REPO.git
+        {ref} is a branch name, tag, or commit SHA
+        {path} is the path to the base of the staging area (e.g. 'tests/area1')
+        {prefix} is the common prefix and partition length (e.g. '/0')
+
+        >>> s = CannedSourceSpec.parse('https://github.com/user/repo.git::f0f0f0f0::dir/subdir::/0')
+        >>> s # doctest: +NORMALIZE_WHITESPACE
+        CannedSourceSpec(prefix=Prefix(common='', partition=0),
+                         url='https://github.com/user/repo.git',
+                         ref='f0f0f0f0',
+                         path=('dir', 'subdir'))
+
+        >>> s = CannedSourceSpec.parse('https://github.com/user/repo.git::f0f0f0f0::::42/0')
+        >>> s # doctest: +NORMALIZE_WHITESPACE
+        CannedSourceSpec(prefix=Prefix(common='42', partition=0),
+                         url='https://github.com/user/repo.git',
+                         ref='f0f0f0f0',
+                         path=())
+
+        >>> s = CannedSourceSpec.parse('https://github.com/foo.git::f0f0f0f0::::/0')
+        Traceback (most recent call last):
+        ...
+        azul.RequirementError: ('Invalid GitHub repo URL', 'https://github.com/foo.git')
+        """
+        url, ref, path, prefix = cls._parse(spec)
+        self = cls(prefix=prefix,
+                   url=url,
+                   ref=ref,
+                   path=path)
+        assert spec == str(self), spec
+        return self
+
+    @classmethod
+    def _parse(cls, spec: str) -> tuple[str, str, tuple[str], Prefix]:
+        require(spec.count('::') == 3, 'Invalid source specification', spec)
```

Comment:
```suggestion
        require(spec.count(cls.sep) == 3, 'Invalid source specification', spec)
```

Along with all other occurrences of "::"